firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/metadata/UserMetadata.java

```diff
@@ -135,6 +138,40 @@ public void setUserId(String identifier) {
     crashlyticsWorkers.diskWrite.submit(this::serializeUserDataIfNeeded);
   }
 
+  /**
+   * Returns a {@link Map<String, String>} containing all the custom keys to attach to the event.
+   * It overrides the values of app level custom keys with the values of event level custom keys if
+   * they're identical, and event keys or values that exceed 1024 characters are truncated.
+   * Combined with app level custom keys, the map is restricted to 64 key value pairs.
+   *
+   * @param eventKeys a {@link Map<String, String>} representing event specific keys.
+   * @return a {@link Map<String, String>} containing all the custom keys to attach to the event.
+   */
+  public Map<String, String> getCustomKeys(Map<String, String> eventKeys) {
+    // In case of empty event keys, preserve existing behavior.
+    if (eventKeys.isEmpty()) return customKeys.getKeys();
+
+    // Otherwise merge the event keys with custom keys as appropriate.
+    Map<String, String> globalKeys = customKeys.getKeys();
+    HashMap<String, String> result = new HashMap<>(globalKeys);
+    for (Map.Entry<String, String> entry : eventKeys.entrySet()) {
+      String sanitizedKey = KeysMap.sanitizeString(entry.getKey(), MAX_ATTRIBUTE_SIZE);
+      if (result.size() < MAX_ATTRIBUTES || result.containsKey(sanitizedKey)) {
+        String sanitizedValue = KeysMap.sanitizeString(entry.getValue(), MAX_ATTRIBUTE_SIZE);
+        result.put(sanitizedKey, sanitizedValue);
+        continue;
+      }
+
+      Logger.getLogger()
```

Comment:
Will this log a line per key over size? Can we make it just log once since it's a warning?

Reply:
Done.